File: code_files/week4/week4_topic1_lumped_biot.py
# Week 4 - Topic 1: Lumped Capacitance & the Biot Number
# Unsteady conduction, "0-D" limit: when internal conduction is much faster than
# surface convection (Bi << 1), the whole body is nearly isothermal and the PDE
# collapses to a first-order ODE:
#   rho*c*V dT/dt = -h*A*(T - Tinf)  ->  (T-Tinf)/(T0-Tinf) = exp(-t/tau) = exp(-Bi*Fo)
#   Bi = h*Lc/k,  Fo = alpha*t/Lc^2,  Lc = V/A,  tau = rho*c*Lc/h
# Also: steady radial profile of a heat-generating wire (lecture),
#   Theta(eta) = (2 + Bi*(1 - eta^2)) / (2 + Bi),  eta = r/R
# shows the physical meaning of Bi: Bi->0 isothermal wire, Bi->inf surface at Tinf.

import math

def lumped_history(Bi, Fo):
    """Dimensionless temperature (T-Tinf)/(T0-Tinf) = exp(-Bi*Fo)."""
    return math.exp(-Bi * Fo)

def wire_profile(eta, Bi):
    """Steady dimensionless profile of a heat-generating wire, Theta=(T-Tinf)/(Tc-Tinf)."""
    return (2.0 + Bi * (1.0 - eta * eta)) / (2.0 + Bi)

if __name__ == "__main__":
    # --- Case A: steel sphere quenched in oil ---
    D = 0.010                       # sphere diameter [m]
    rho, c, k = 7800.0, 480.0, 45.0 # steel [kg/m^3, J/kgK, W/mK]
    h = 400.0                       # oil bath [W/m^2K]
    T0, Tinf = 850.0, 60.0          # initial & bath temperature [degC]

    Lc = D / 6.0                    # V/A of a sphere = D/6
    Bi = h * Lc / k
    alpha = k / (rho * c)
    tau = rho * c * Lc / h
    print("Case A: steel sphere, D = 10 mm, quenched in oil")
    print(f"  Lc = V/A = D/6 = {Lc*1e3:.3f} mm")
    print(f"  Bi = h*Lc/k = {Bi:.4f}  ({'< 0.1 -> lumped OK' if Bi < 0.1 else '>= 0.1 -> need full PDE'})")
    print(f"  alpha = {alpha*1e6:.3f} mm^2/s,  tau = rho*c*Lc/h = {tau:.2f} s")

    print("  t [s]   Fo      T [degC]")
    for t in (0.0, 5.0, 15.6, 30.0, 60.0, 120.0):
        Fo = alpha * t / Lc**2
        theta = lumped_history(Bi, Fo)          # = exp(-t/tau)
        T = Tinf + (T0 - Tinf) * theta
        print(f"  {t:5.1f}  {Fo:6.2f}  {T:8.2f}")
    t99 = tau * math.log(100.0)
    print(f"  time to 99% equilibration: t = tau*ln(100) = {t99:.1f} s")

    # sanity: exp(-Bi*Fo) must equal exp(-t/tau)
    t = 37.0
    Fo = alpha * t / Lc**2
    assert abs(lumped_history(Bi, Fo) - math.exp(-t / tau)) < 1e-12

    # --- Case B: physical meaning of Bi via the heat-generating wire (lecture) ---
    print("\nCase B: steady heat-generating wire, Theta = (2 + Bi(1-eta^2))/(2 + Bi)")
    print("  eta:   " + "  ".join(f"{e:5.2f}" for e in (0.0, 0.25, 0.5, 0.75, 1.0)))
    for Bi_w in (0.0, 0.5, 2.0, 10.0, 1e12):
        vals = [wire_profile(e, Bi_w) for e in (0.0, 0.25, 0.5, 0.75, 1.0)]
        tag = "inf" if Bi_w > 1e6 else f"{Bi_w:g}"
        print(f"  Bi={tag:>4s}: " + "  ".join(f"{v:5.3f}" for v in vals))
    print("  Bi->0  : Theta = 1 everywhere (isothermal wire, convection-limited)")
    print("  Bi->inf: Theta(1) = 0 (surface pinned at Tinf, conduction-limited)")

File: code_files/week4/week4_topic1_lumped_biot.jl
# Week 4 - Topic 1: Lumped Capacitance & the Biot Number
# (T-Tinf)/(T0-Tinf) = exp(-t/tau) = exp(-Bi*Fo), Bi = h*Lc/k, Fo = alpha*t/Lc^2
# Also the steady heat-generating wire (lecture): Theta = (2 + Bi*(1-eta^2))/(2 + Bi)
# Run: julia week4_topic1_lumped_biot.jl

using Printf

lumped_history(Bi, Fo) = exp(-Bi * Fo)
wire_profile(eta, Bi) = (2 + Bi * (1 - eta^2)) / (2 + Bi)

# --- Case A: steel sphere quenched in oil ---
D = 0.010
rho, c, k = 7800.0, 480.0, 45.0
h = 400.0
T0, Tinf = 850.0, 60.0

Lc = D / 6                       # V/A of a sphere
Bi = h * Lc / k
alpha = k / (rho * c)
tau = rho * c * Lc / h

println("Case A: steel sphere, D = 10 mm, quenched in oil")
@printf("  Lc = D/6 = %.3f mm\n", Lc * 1e3)
@printf("  Bi = %.4f  (%s)\n", Bi, Bi < 0.1 ? "< 0.1 -> lumped OK" : ">= 0.1 -> full PDE")
@printf("  alpha = %.3f mm^2/s,  tau = %.2f s\n", alpha * 1e6, tau)

println("  t [s]   Fo      T [degC]")
for t in (0.0, 5.0, 15.6, 30.0, 60.0, 120.0)
    Fo = alpha * t / Lc^2
    T = Tinf + (T0 - Tinf) * lumped_history(Bi, Fo)   # = exp(-t/tau)
    @printf("  %5.1f  %6.2f  %8.2f\n", t, Fo, T)
end
@printf("  time to 99%% equilibration: %.1f s\n", tau * log(100))

# sanity: exp(-Bi*Fo) must equal exp(-t/tau)
let t = 37.0
    Fo = alpha * t / Lc^2
    @assert abs(lumped_history(Bi, Fo) - exp(-t / tau)) < 1e-12
end

# --- Case B: physical meaning of Bi via the heat-generating wire ---
println("\nCase B: steady wire, Theta = (2 + Bi(1-eta^2))/(2 + Bi)")
etas = (0.0, 0.25, 0.5, 0.75, 1.0)
println("  eta:   " * join([@sprintf("%5.2f", e) for e in etas], "  "))
for Biw in (0.0, 0.5, 2.0, 10.0, 1e12)
    tag = Biw > 1e6 ? " inf" : @sprintf("%4g", Biw)
    vals = join([@sprintf("%5.3f", wire_profile(e, Biw)) for e in etas], "  ")
    println("  Bi=$tag: $vals")
end
println("  Bi->0: isothermal wire (convection-limited)")
println("  Bi->inf: surface pinned at Tinf (conduction-limited)")
